CorsHandler to respect http connection (keep-alive) header.

willblackie opened the pull request:
Motivation:

The CorsHandler currently closes the channel when it responds to a preflight (OPTIONS)
request or in the event of a short circuit due to failed validation.

Especially in an environment where there's a proxy in front of the service this causes
unnecessary connection churn.

Modifications:

CorsHandler now uses HttpUtil to determine if the connection should be closed
after responding

Result:

Channel will stay open when the CorsHandler responds unless the client specifies otherwise
or the protocol version is HTTP/1.0

normanmaurer commented:
@willblackie please fix the check style error:
https://garage.netty.io/teamcity/viewLog.html?buildId=11769&buildTypeId=pulls_netty&guest=1

willblackie commented:
think i have it sorted now, please look again.

willblackie commented:
@normanmaurer anything else i need to do on this? thanks

normanmaurer commented:
Will check shortly...

Did you sign the icla yet?
http://netty.io/s/icla



> Am 23.08.2016 um 19:18 schrieb willblackie <notifications@github.com>:
> 
> @normanmaurer anything else i need to do on this? thanks
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

willblackie commented:
yes icla done, I'm not in a huge hurry just making sure there's nothing else I need to do (it's also been through a decent amount of load testing already).. thanks

normanmaurer commented on codec-http/src/main/java/io/netty/handler/codec/http/cors/CorsHandler.java:
```diff
@@ -80,7 +81,12 @@ private void handlePreflight(final ChannelHandlerContext ctx, final HttpRequest
+            ctx.writeAndFlush(response);
+        } else {
+            ctx.writeAndFlush(response).addListener(ChannelFutureListener.CLOSE);
+        }
```
Just use:

```
ChannelFuture future = ctx.writeAndFlush(response);
if (!HttpUtil.isKeepAlive(request)) {
    future.addListener(ChannelFutureListener.CLOSE);
}
```

normanmaurer commented on codec-http/src/main/java/io/netty/handler/codec/http/cors/CorsHandler.java:
```diff
@@ -203,8 +209,14 @@ public void write(final ChannelHandlerContext ctx, final Object msg, final Chann
+        } else {
+            ctx.writeAndFlush(new DefaultFullHttpResponse(request.protocolVersion(), FORBIDDEN))
+                    .addListener(ChannelFutureListener.CLOSE);
+        }
```
see above

normanmaurer commented on codec-http/src/test/java/io/netty/handler/codec/http/cors/CorsHandlerTest.java:
```diff
@@ -289,17 +293,100 @@ public void shortCurcuitNonCorsRequest() {
+        request.headers().set(ORIGIN, "http://localhost:8888");
+        request.headers().set(CONNECTION, KEEP_ALIVE);
+
+        channel.writeInbound(request);
```
assert return value.

normanmaurer replied:
@willblackie please address the comment this will either return true or false. So assert it ;)

willblackie replied:
@normanmaurer done

normanmaurer commented on codec-http/src/test/java/io/netty/handler/codec/http/cors/CorsHandlerTest.java:
```diff
@@ -289,17 +293,100 @@ public void shortCurcuitNonCorsRequest() {
+        final HttpResponse response = channel.readOutbound();
+
+        assertThat(channel.isOpen(), is(true));
+        assertThat(response.status(), is(FORBIDDEN));
```
also call `channel.finish()` and assert return value and ensure you release all the stuff in the `EmbeddedChannel`

normanmaurer commented on codec-http/src/test/java/io/netty/handler/codec/http/cors/CorsHandlerTest.java:
```diff
@@ -289,17 +293,100 @@ public void shortCurcuitNonCorsRequest() {
+        final HttpResponse response = channel.readOutbound();
+
+        assertThat(channel.isOpen(), is(true));
+        assertThat(response.status(), is(FORBIDDEN));
```
see above

normanmaurer commented on codec-http/src/test/java/io/netty/handler/codec/http/cors/CorsHandlerTest.java:
```diff
@@ -289,17 +293,100 @@ public void shortCurcuitNonCorsRequest() {
+        final HttpResponse response = channel.readOutbound();
+
+        assertThat(channel.isOpen(), is(false));
+        assertThat(response.status(), is(FORBIDDEN));
```
see above

normanmaurer commented on codec-http/src/test/java/io/netty/handler/codec/http/cors/CorsHandlerTest.java:
```diff
@@ -289,17 +293,100 @@ public void shortCurcuitNonCorsRequest() {
+        final HttpResponse response = channel.readOutbound();
+
+        assertThat(channel.isOpen(), is(true));
+        assertThat(response.status(), is(OK));
```
see above

normanmaurer commented on codec-http/src/test/java/io/netty/handler/codec/http/cors/CorsHandlerTest.java:
```diff
@@ -289,17 +293,100 @@ public void shortCurcuitNonCorsRequest() {
+        final HttpResponse response = channel.readOutbound();
+
+        assertThat(channel.isOpen(), is(true));
+        assertThat(response.status(), is(OK));
```
see above

normanmaurer commented on codec-http/src/test/java/io/netty/handler/codec/http/cors/CorsHandlerTest.java:
```diff
@@ -289,17 +293,100 @@ public void shortCurcuitNonCorsRequest() {
+        final HttpResponse response = channel.readOutbound();
+
+        assertThat(channel.isOpen(), is(false));
+        assertThat(response.status(), is(OK));
```
see above

normanmaurer commented on codec-http/src/test/java/io/netty/handler/codec/http/cors/CorsHandlerTest.java:
```diff
@@ -339,15 +426,22 @@ private static HttpResponse preflightRequest(final CorsConfig config,
+
+        if (connection != null) {
+            httpRequest.headers().set(CONNECTION, connection);
+        }
```
see above

normanmaurer commented:
@willblackie please address comments

willblackie commented:
@normanmaurer thanks for the review. i've addressed most of your comments and the build has passed, just not sure what you want me to do with the code below.  i don't love the AsciiString argument but needed a way to test for 3 variations (Connection: Close, Connection: Keep-Alive & absence of the header) if you have something that's more in keeping with the Netty codebase style i'm happy to change it.

```java
if (connection != null) {
    httpRequest.headers().set(CONNECTION, connection);
}
```

willblackie commented:
@normanmaurer please take a look .. thanks

otoolec commented on codec-http/src/main/java/io/netty/handler/codec/http/cors/CorsHandler.java:
```diff
@@ -80,7 +82,11 @@ private void handlePreflight(final ChannelHandlerContext ctx, final HttpRequest
         release(request);
-        ctx.writeAndFlush(response).addListener(ChannelFutureListener.CLOSE);
+
+        final ChannelFuture future = ctx.writeAndFlush(response);
```
If the protocol version is 1.0 and the request was marked ask keep-alive, then shouldn't the response be marked as keep-alive?  Do you need something like this here?

```java
if (HttpUtil.isKeepAlive(request)) {
  HttpUtil.setKeepAlive(response, true);
}
```

willblackie replied:
good point, I've moved both places in the CorsHandler that respond into a separate method that does the keep alive check and also sets the keep alive on the response.  There's no harm in always setting it right?  I can omit it in the case where it's 1.0 and not specified in the request if necessary.

@normanmaurer just pinging you to make sure you see this on your next review

netkins commented on codec-http/src/main/java/io/netty/handler/codec/http/cors/CorsHandler.java:
```diff
@@ -16,14 +16,17 @@
 import io.netty.channel.ChannelHandlerContext;
 import io.netty.channel.ChannelPromise;
 import io.netty.handler.codec.http.DefaultFullHttpResponse;
+import io.netty.handler.codec.http.FullHttpResponse;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.handler.codec.http.FullHttpResponse'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

willblackie commented:
@normanmaurer is the SonarQube fail my fault?  If so is there a way I can see the results in order to fix?

normanmaurer commented:
LGTM... @otoolec wdyt ?

otoolec commented:
I think the latest code will work and handles the concern I had. 

@willblackie brings up a good point about where the Keep Alive logic should be.  It seems obvious to have such logic at the point in code where a request is read and a response is generated, since we need access to both for this logic.  It would be more convenient to have a handler take care of this for developers though.

I like the idea of having the HttpKeepAliveHandler and if we added it to the HttpServerCodec then it would even solve this same problem for HttpOjbectAggregator [when it responds with TOO_LARGE](https://github.com/netty/netty/blob/30fe2e868fcc99d24046e505f9406067ba6e9d07/codec-http/src/main/java/io/netty/handler/codec/http/HttpObjectAggregator.java#L219) on a non-persistent connection without additional changes to the current recommendations.  

I wonder if adding HttpKeepAliveHandler would lead to some unexpected behavior for existing users.  Maybe that can be considered for 5.x but avoided for 4.x.

willblackie commented:
To address the backwards compatibility issue HttpServerCodec could have another constructor that takes an "autoKeepAlive" parameter which defaults to false (current behavior) then when true it includes the HttpKeepAliveHandler.  Also the default could move to true in 5.x .. it makes sense (to me) that this is handled properly by default.

otoolec commented:
@willblackie take a look at #5788 and let me know what you think.  My thinking is if that PR gets accepted then there could be a follow on PR to include the handler in HttpServerCodec.  Would you be interested in working on that follow on PR since it sounds like you had some good ideas on how to handle the default behavior?

willblackie commented:
@otoolec yeah I'm happy to work on it, given that do you think I modify this PR to just never call close on the channel? .. could follow with a similar fix for the HttpObjectAggregator as you pointed out earlier (and do a search for others).

otoolec commented:
@willblackie I think the decision of what changes need to be made to this PR are up to @normanmaurer.  If the Netty project wants to go in a different direction (not use HttpKeepAliveHandler) then your code should probably stay as is; if HttpKeepAliveHandler is added then you may want to reference it in the class docs as part of this PR.

HttpObjectAggregator's listener should only be doing a close on failure, basically a `CLOSE_ON_FAILURE` with a log statement, so it can probably be kept since we would want to close the connection on failure even if it was a persistent connection.  I guess that brings up the question of if the HttpKeepAliveHandler should be going as far as adding the `CLOSE_ON_FAILURE` listener as well.  I don't think I have a good enough understanding of Netty to say if that makes sense.

normanmaurer commented:
@otoolec @willblackie I think we should keep this PR as it is and I will check the other PR with the `HttpKeepAliveHandler` separately.

normanmaurer commented:
Cherry-picked into 4.1 (e3aca1f3d6f8ee2a2c016b65fab19728eb01bc6f) and 4.0 (3a4403bb554c7c1dbbb39eef005dfbf67fd95ae8)

willblackie commented:
thanks @normanmaurer
